### PR TITLE
nixos/test-driver: fix #153766

### LINKED ISSUE #153766: nixos/test-driver: "bare" driver crashes when $out env var is not a writable directory
### Steps To Reproduce
1. `nix-shell -p hello`
2. `nix-build . -A "nixosTests.sway.driver"
3. `./result/bin/nixos-test-driver`

### Expected behavior
Driver runs normally, placing outputs in current directory.

### Driver output:

```
machine: waiting for file ‘/tmp/test-x11-exit-ok‘
(finished: waiting for file ‘/tmp/test-x11-exit-ok‘, in 1.12 seconds)
machine: must succeed: cat /tmp/test-x11.out
(finished: must succeed: cat /tmp/test-x11.out, in 0.03 seconds)
GL_VERSION: 4.5 (Compatibility Profile) Mesa 21.3.2
GL_RENDERER: llvmpipe (LLVM 13.0.0, 128 bits)
GL_VENDOR: Mesa/X.org

machine: making screenshot /nix/store/99k3djmxmgynmsdcpd1qfy953x07k5pp-shell/alacritty_glinfo.png
machine: sending monitor command: screendump /nix/store/99k3djmxmgynmsdcpd1qfy953x07k5pp-shell/alacritty_glinfo.png.ppm
machine: waiting for monitor prompt
(finished: waiting for monitor prompt, in 0.00 seconds)
(finished: sending monitor command: screendump /nix/store/99k3djmxmgynmsdcpd1qfy953x07k5pp-shell/alacritty_glinfo.png.ppm, in 0.00 seconds)
/nix/store/iqprjr5k5385bhf1dzj07zwd5p43py1n-bash-5.1-p12/bin/sh: line 1: /nix/store/99k3djmxmgynmsdcpd1qfy953x07k5pp-shell/alacritty_glinfo.png: No such file or directory
cleanup
kill machine (pid 592567)
machine # qemu-system-x86_64: terminating on signal 15 from pid 592565 (/nix/store/rppr9s436950i1dlzknbmz40m2xqqnxc-python3-3.9.9/bin/python3.9)
(finished: cleanup, in 0.02 seconds)
Traceback (most recent call last):
  File "/nix/store/zmyxgfrr2lhfcgczljwrh8h50jhs06lx-nixos-test-driver-1.0/bin/.nixos-test-driver-wrapped", line 9, in <module>
    sys.exit(main())
  File "/nix/store/zmyxgfrr2lhfcgczljwrh8h50jhs06lx-nixos-test-driver-1.0/lib/python3.9/site-packages/test_driver/__init__.py", line 86, in main
    driver.run_tests()
  File "/nix/store/zmyxgfrr2lhfcgczljwrh8h50jhs06lx-nixos-test-driver-1.0/lib/python3.9/site-packages/test_driver/driver.py", line 115, in run_tests
    self.test_script()
  File "/nix/store/zmyxgfrr2lhfcgczljwrh8h50jhs06lx-nixos-test-driver-1.0/lib/python3.9/site-packages/test_driver/driver.py", line 111, in test_script
    exec(self.tests, symbols, None)
  File "<string>", line 19, in <module>
  File "/nix/store/zmyxgfrr2lhfcgczljwrh8h50jhs06lx-nixos-test-driver-1.0/lib/python3.9/site-packages/test_driver/machine.py", line 712, in screenshot
    os.unlink(tmp)
FileNotFoundError: [Errno 2] No such file or directory: '/nix/store/99k3djmxmgynmsdcpd1qfy953x07k5pp-shell/alacritty_glinfo.png.ppm'
kill vlan (pid 592566)
```

### Notify maintainers

... Now that we're packaging `test-driver` as a proper package we should probably also include a meta section

Anyways: @marijanp @Artturin @cole-h

### PR DESCRIPTION
**Pull request by Synthetica9**
###### Motivation for this change

Fixes #153766, alternative to #153854

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/lib/test-driver/test_driver/machine.py [hunk 1]
```diff
@@ -336,6 +338,10 @@ def __init__(
         self.name = name
         self.start_command = start_command
 
+        if out_dir is None:
+            out_dir = Path(".").resolve()
```
Comment (marijanp) [inline nixos/lib/test-driver/test_driver/machine.py:342]: I also thought about making the out-dir an optional parameter, defaulting to the PWD. Now the problem I thought of was, if a user runs the driver inside a directory without write-permission. The test-run will fail pretty late i.e. when `screenshot` or `copy_from_vm` is called. I found it nicer to fail as early as possible. The case can occur e.g. when someone does this:
```
nix-build . -A nixosTests.<test>.driver
cd result/bin
./nixos-test-driver
```
Here the result dir is a symlink to the nix store obviously...

Comment (Synthetica9) [reply nixos/lib/test-driver/test_driver/machine.py]: agree, but I think that check should be happening in the initialization of Driver, not in parsing the arguments. I rebased and pushed a new commit to fix this, please verify

**Comment by marijanp** [comment]
Can we get rid of all accesses to environment variables, I already did that in https://github.com/NixOS/nixpkgs/pull/153854/files
And you will also need @tfc s approval as he's the code owner

**Comment by tfc** [comment]
Hi, i tend to prefer the other PR, because it is already cleaned out from env var usage.

**Comment by Synthetica9** [comment]
I have some thoughts on the other one, but I'll leave them there.

**Pull request closed without merging**